### PR TITLE
Remove unused LvmPv when editing a partition

### PR DESCRIPTION
**Pull request by dgdavid**
## Problem

Editing a partition with LvmPv not associated with any LvmVg could generate not doable operations in the real system. More specifically, 

> If a partition contains an LvmPv not associated to any LvmVg, it's possible to edit the partition and mark the "encrypt device" check-box. The result is a devicegraph in which a new Encrypt device in inserted between the partition and the existing LvmPv

- https://bugzilla.suse.com/show_bug.cgi?id=1129663
- https://trello.com/c/RMJphyX3/914-tumbleweed-1129663-setting-up-encrypted-partition-with-yast-errors-stub-doreallottext-called-for-lvmpv

## Solution

Remove unused LvmPv when a partition is edited.

## Testing

- Added a new unit test
- Also Tested manually

## Screenshots

Given the [following devicegraph](https://github.com/yast/yast-storage-ng/files/3062824/empty_lvm_pv.zip)

original_device_graph

### :bug:  Results with current code base

| After going through the /dev/sda2 edit dialog | After add encryption to /dev/sda2 |
|-----------------------------------|----------------------------------------|
| after_going_through_edit_dialog_sda3_wo_changes_before_fix png_sda2_wo_fix | after_donotformat_and_encrypt_sda2_wo_fix |

| After keep the /dev/sda3 encryption | After remove the /dev/sda3 encryption |
|-----------------------------------|----------------------------------------|
| after_going_through_edit_dialog_sda3_wo_changes_before_fix png_sda2_wo_fix | after_remove_encryption_sda3_before_fix |

### :heart: Results after merge proposed changes
| After going through the /dev/sda2 edit dialog | After add encryption to /dev/sda2 |
|-----------------------------------|----------------------------------------|
| after_going_through_edit_dialog_sda2_wo_changes |  after_donotformat_and-encrypt_sda2 |

| After keep the /dev/sda3 encryption | After remove the /dev/sda3 encryption |
|-----------------------------------|----------------------------------------|
| after_going_through_edit_dialog_sda3_wo_changes | after_remove_encryption_sda3_wo_format |

---

[empty_lvm_pv.zip](https://github.com/yast/yast-storage-ng/files/3062824/empty_lvm_pv.zip)

### CODE REVIEW

File: package/yast2-storage-ng.changes [hunk 1]
```diff
@@ -1,3 +1,11 @@
+-------------------------------------------------------------------
+Wed Apr 10 07:55:26 UTC 2019 - David Diaz <dgonzalez@suse.com>
+
+- Fix partitioner to not have not doable operations after editing
+  a partition that initially contains an LvmPv not associated with
+  any LvmVg (bsc#1129663).
```
Comment (joseivanlopez) [inline package/yast2-storage-ng.changes]: NP: I find the explanation a little bit confusing. Maybe something like the following?

Partitioner: when editing a partition, remove its orphan LVM Physical Volume. This avoid no doable operations (bsc#1129663).

File: src/lib/y2partitioner/actions/controllers/filesystem.rb [hunk 2]
```diff
@@ -344,6 +348,17 @@ def finish
...
 
+        # Removes from the block device or its encryption layer a LvmPv not associated to an LvmVg
+        # (bsc#1129663)
+        def remove_unused_lvm_pv
+          device = blk_device.encryption ? blk_device.encryption : blk_device
+          lvm_pv = device.lvm_pv
+
+          return unless device.lvm_pv
+
+          device.remove_descendants if lvm_pv.descendants.empty?
```
Comment (joseivanlopez) [inline src/lib/y2partitioner/actions/controllers/filesystem.rb]: NP:
```suggestion
          device = blk_device.encryption || blk_device
          lvm_pv = device.lvm_pv

          device.remove_descendants if lvm_pv && lvm_pv.descendants.none?
```

File: package/yast2-storage-ng.changes (hunk 1 above)
Comment (ancorgs) [inline package/yast2-storage-ng.changes]: Not that easy to parse for a user. I would go for something like this.

```
- Partitioner: when editing a block device, clean-up useless LVM PV metadata
  from it (bsc#1129663)
```

File: src/lib/y2partitioner/actions/controllers/filesystem.rb (hunk 2 above)
Comment (dgdavid) [reply src/lib/y2partitioner/actions/controllers/filesystem.rb]: > evice = blk_device.encryption || blk_device

Nice. Believe or not, I was thinking for a while: "David, there is something wrong with that assignation using a ternary operator"

But then I forgot to change it. 

Thanks!

File: package/yast2-storage-ng.changes (hunk 1 above)
Comment (dgdavid) [reply package/yast2-storage-ng.changes]: >  I find the explanation a little bit confusing. 

I agree. However, I'll use the [Ancor's suggestion](https://github.com/yast/yast-storage-ng/pull/897#pullrequestreview-224884529), if you do not mind :wink:

File: src/lib/y2partitioner/actions/controllers/filesystem.rb [hunk 3]
```diff
@@ -356,7 +356,10 @@ def remove_unused_lvm_pv
 
           return unless device.lvm_pv
 
-          device.remove_descendants if lvm_pv.descendants.empty?
+          device = blk_device.encryption || blk_device
+          lvm_pv = device.lvm_pv
+
+          device.remove_descendants if lvm_pv && lvm_pv.descendants.none?
```
Comment (joseivanlopez) [inline src/lib/y2partitioner/actions/controllers/filesystem.rb]: Be careful, not the method is defined as:

```
        def remove_unused_lvm_pv
          device = blk_device.encryption ? blk_device.encryption : blk_device
          lvm_pv = device.lvm_pv

          return unless device.lvm_pv

          device = blk_device.encryption || blk_device
          lvm_pv = device.lvm_pv

          device.remove_descendants if lvm_pv && lvm_pv.descendants.none?
        end
```

But my proposal was:

```
        def remove_unused_lvm_pv
          device = blk_device.encryption || blk_device
          lvm_pv = device.lvm_pv

          device.remove_descendants if lvm_pv && lvm_pv.descendants.none?
        end
``` 

Sorry, maybe my suggestion was misleading.

Comment (dgdavid) [reply src/lib/y2partitioner/actions/controllers/filesystem.rb]: No worries, I'll fix it.

**Pull request merged**